sdk/dapp-kit/src/components/AutoConnectWallet.tsx

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { ReactNode } from 'react';
+import { useEffect, useState } from 'react';
+import { useConnectWallet } from '../hooks/wallet/useConnectWallet.js';
+import { useWalletContext } from '../hooks/wallet/useWalletContext.js';
+import type { StorageAdapter } from '../utils/storageAdapters.js';
+import type { WalletAccount, WalletWithRequiredFeatures } from '@mysten/wallet-standard';
+
+type AutoConnectWalletProps = {
+	storageKey: string;
+	storageAdapter: StorageAdapter;
+	children: ReactNode;
+};
+
+export function AutoConnectWallet({
+	storageKey,
+	storageAdapter,
+	children,
+}: AutoConnectWalletProps) {
+	const { wallets, currentWallet: wallet, currentAccount: account } = useWalletContext();
+	const [previousWallet, setPreviousWallet] = useState<WalletWithRequiredFeatures | null>(null);
+	const [currentWallet, setCurrentWallet] = useState<WalletWithRequiredFeatures | null>(wallet);
+	const [previousAccount, setPreviousAccount] = useState<WalletAccount | null>(null);
+	const [currentAccount, setCurrentAccount] = useState<WalletAccount | null>(account);
+	const { mutate: connectWallet } = useConnectWallet();
+
+	// Instead of abstracting the previous wallet and account state logic into a generic
```

Comment:
I already know you're gonna hit me with a wtf on this haha, I decided to forgo abstracting this into `usePrevious` since IMO it's not a pattern we want to normally encourage + the shallow re-rendering stuff does add a little complexity that might not be good to hide. If you have a strong preference, I'm glad to go the `usePrevious` route!

https://twitter.com/dan_abramov/status/1653422709508849664